collectors/diskspace.plugin/plugin_diskspace.c

```diff
@@ -106,10 +123,208 @@ int mount_point_is_protected(char *mount_point)
     return 0;
 }
 
-static inline void do_disk_space_stats(struct mountinfo *mi, int update_every) {
+// a copy of basic mountinfo fields
+struct basic_mountinfo {
+    char *persistent_id;    
+    char *root;             
+    char *mount_point;      
+    char *filesystem;       
+
+    struct basic_mountinfo *next;
+};
+
+static struct basic_mountinfo *slow_mountinfo_tmp_root = NULL;
+static netdata_mutex_t slow_mountinfo_mutex;
+
+static struct basic_mountinfo *basic_mountinfo_create_and_copy(struct mountinfo* mi)
+{
+    struct basic_mountinfo *bmi = callocz(1, sizeof(struct basic_mountinfo));
+    
+    if (mi) {
+        bmi->persistent_id = strdupz(mi->persistent_id);
+        bmi->root          = strdupz(mi->root);
+        bmi->mount_point   = strdupz(mi->mount_point);
+        bmi->filesystem    = strdupz(mi->filesystem);
+    }
+
+    return bmi;
+}
+
+static void add_basic_mountinfo(struct basic_mountinfo **root, struct mountinfo *mi)
+{
+    if (!root)
+        return;
+
+    struct basic_mountinfo *bmi = basic_mountinfo_create_and_copy(mi);
+
+    bmi->next = *root;
+    *root = bmi;
+};
+
+static void free_basic_mountinfo(struct basic_mountinfo *bmi)
+{
+    if (bmi) {
+        freez(bmi->persistent_id);
+        freez(bmi->root);
+        freez(bmi->mount_point);
+        freez(bmi->filesystem);
+    }
+
+    freez(bmi);
+};
+
+static void free_basic_mountinfo_list(struct basic_mountinfo *root)
+{
+    struct basic_mountinfo *bmi = root, *next;
+
+    while (bmi) {
+        next = bmi->next;
+        free_basic_mountinfo(bmi);
+        bmi = next;
+    }
+}
+
+static void calculate_values_and_show_charts(
+    struct basic_mountinfo *mi,
+    struct mount_point_metadata *m,
+    struct statvfs *buff_statvfs,
+    int update_every)
+{
     const char *family = mi->mount_point;
     const char *disk = mi->persistent_id;
 
+    // logic found at get_fs_usage() in coreutils
+    unsigned long bsize = (buff_statvfs->f_frsize) ? buff_statvfs->f_frsize : buff_statvfs->f_bsize;
+
+    fsblkcnt_t bavail         = buff_statvfs->f_bavail;
+    fsblkcnt_t btotal         = buff_statvfs->f_blocks;
+    fsblkcnt_t bavail_root    = buff_statvfs->f_bfree;
+    fsblkcnt_t breserved_root = bavail_root - bavail;
+    fsblkcnt_t bused;
```

Comment:
We could convert the four lines a to a ternary.

Reply:
Hm 🤔, what do you mean? Could you elaborate a bit?

Reply:
It is a minor suggestion, I was saying that next lines can be rewritten as:

```c
bused = (likely(btotal >= bavail_root)) ? btotal - bavail_root : bavail_root - btotal;
```

Reply:
Ah, got it 😄. You are talking about the next four lines, not the displayed ones.

Reply:
yes :) .